base/inc/AdePT/TrackManager.h

```diff
@@ -0,0 +1,263 @@
+ *          slots for the current and next iteration.
+ *
+ *          The compression of the sparse active slots is triggered when the sum of the
+ *          number of used slots and 3 times the number of inflight tracks becomes larger
+ *          than a user-defined fraction of the total buffer size. The tracks pointed by the
+ *          next iteration array are copied cotiguously by a kernel just after the last used
+ *          slot, before making this slot the start index and swapping the current/next index
+ *          arrays.
+ *
+ * @author Andrei Gheata (andrei.gheata@cern.ch)
+ */
+
+#ifndef ADEPT_TRACKMANAGER_H_
+#define ADEPT_TRACKMANAGER_H_
+
+#include <type_traits>
+#include <CopCore/CopCore.h>
+#include <AdePT/Atomic.h>
+
+namespace adept {
+
+template <typename Track>
+struct TrackManager;
+
+#ifdef COPCORE_CUDA_COMPILER
+namespace device_impl_trackmgr {
+
+template <typename Track>
+__global__ void construct_trackmanager(void *addr, size_t capacity, adept::MParray *activeSlots,
+                                       adept::MParray *nextSlots, Track *buffer)
+{
+  // Invoke inplace TrackManager constructor
+  auto mgr           = new (addr) TrackManager<Track>(capacity);
+  mgr->fActiveTracks = activeSlots;
+  mgr->fNextTracks   = nextSlots;
+  mgr->fBuffer       = buffer;
+  // construct the MParray objects inplace
+  adept::MParray::MakeInstanceAt(capacity, activeSlots);
+  adept::MParray::MakeInstanceAt(capacity, nextSlots);
+}
+
+template <typename Track>
+__global__ void clear_trackmanager(TrackManager<Track> *mgr)
+{
+  mgr->clear();
+}
+
+template <typename Track>
+__global__ void swap_active(TrackManager<Track> *mgr)
+{
+  mgr->swap();
+}
+
+template <typename Track>
+__global__ void defragment_buffer(TrackManager<Track> *mgr, int nactive, int where)
+{
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < nactive; i += blockDim.x * gridDim.x) {
+    const int slot_src     = (*mgr->fNextTracks)[i];
+    const int slot_dst     = (where + i) & (mgr->fCapacity - 1);
+    mgr->fBuffer[slot_dst] = mgr->fBuffer[slot_src];
+    // fActiveTracks must be cleared before starting this kernel
+    mgr->fActiveTracks->push_back(slot_dst);
+  }
+}
+
+template <typename Track>
+__global__ void adjust_indices(TrackManager<Track> *mgr, int start_new, int next_free)
+{
+  mgr->fStats.fStart = start_new;
+  mgr->fNextFree.store(next_free);
+  mgr->fNextTracks->clear();
+}
+
+} // End namespace device_impl_trackmgr
+#endif
+
+/// @brief A track manager working with a circular buffer.
+template <typename Track>
+struct TrackManager {
+
+  /// @brief Statistics of the track manager to be updated and copied to host between iterations
+  struct Stats {
+    int fStart{0};     ///< Index of first used track in the buffer
+    int fNextStart{0}; ///< Index where the tracks will be compacted
+    int fInFlight{0};  ///< Number of tracks sill in flight
+
+    int GetNused() { return fNextStart - fStart; }
```

Comment:
Does this work / need to work if the buffer wraps around? I'd expect some `+capacity` and `%capacity` going on here...

Reply:
Yes, fNextStart is always larger than fStart, it may exceed the buffer size. Slot numbers are systematically adjusted to modulo fCapacity